File: KCD/HMNyukyoSpeedChangeCommand.m
//
//  HMNyukyoSpeedChangeCommand.m
//  KCD
//
//  Created by Hori,Masaki on 2014/07/21.
//  Copyright (c) 2014年 Hori,Masaki. All rights reserved.
//

#import "HMNyukyoSpeedChangeCommand.h"

#import "HMServerDataStore.h"
#import "HMKCNyukyoDock.h"
#import "HMKCShipObject+Extensions.h"


@implementation HMNyukyoSpeedChangeCommand
+ (void)load
{
	static dispatch_once_t onceToken;
	dispatch_once(&onceToken, ^{
		[HMJSONCommand registerClass:self];
	});
}

+ (BOOL)canExcuteAPI:(NSString *)api
{
	if([api isEqualToString:@"/kcsapi/api_req_nyukyo/speedchange"]) return YES;
	
	return NO;
}
- (void)execute
{
	HMServerDataStore *store = [HMServerDataStore oneTimeEditor];
	
	NSString *ndockId = self.arguments[@"api_ndock_id"];
	
	NSError *error = nil;
	NSArray<HMKCNyukyoDock *> *nyukyoDocks = [store objectsWithEntityName:@"NyukyoDock"
															  error:&error
														  predicateFormat:@"id = %@", @([ndockId integerValue])];
	if(nyukyoDocks.count == 0) {
		if(error) {
			NSLog(@"Error: at %@ : %@", NSStringFromClass([self class]), error);
		}
		return;
	}
	
	NSNumber *shipId = nyukyoDocks[0].ship_id;
	
	nyukyoDocks[0].ship_id = nil;
	nyukyoDocks[0].state = @(0);
	
	// 艦隊リスト更新用
	error = nil;
	NSArray<HMKCShipObject *> *ships = [store objectsWithEntityName:@"Ship"
															  error:&error
													predicateFormat:@"id = %@", @([shipId integerValue])];
	if(ships.count == 0) {
		if(error) {
			NSLog(@"Error: at %@ : %@", NSStringFromClass([self class]), error);
		}
		return;
	}
	
	ships[0].nowhp = ships[0].maxhp;
}
@end

File: KCD/HMNyukyoSpeedChangeCommand.swift
//
//  HMNyukyoSpeedChangeCommand.swift
//  KCD
//
//  Created by Hori,Masaki on 2015/01/20.
//  Copyright (c) 2015年 Hori,Masaki. All rights reserved.
//

import Cocoa

class HMNyukyoSpeedChangeCommand: HMJSONCommand {
	required override init() {}
	
	override class func load() {
		HMJSONCommand.registerInstance(self())
	}
	override class func canExcuteAPI(api: String) -> Bool {
		return api == "/kcsapi/api_req_nyukyo/speedchange"
	}
	
	override func execute() {
		if let ndockID = arguments["api_ndock_id"] as? String {
			let store = HMServerDataStore.oneTimeEditor()
			let predicate = NSPredicate(format: "id = \(ndockID)")
			var error: NSError? = nil
			let array = store.objectsWithEntityName("NyukyoDock",
				sortDescriptors: nil,
				predicate: predicate,
				error: &error)
			if array.count == 0 {
				if error != nil {
					log("error -> \(error!)", argList: getVaList([]))
				}
				return
			}
			
			if let dock = array[0] as? NSManagedObject {
				dock.setValue(nil, forKey: "ship_id")
				dock.setValue(NSNumber(long: 0), forKey: "state")
				
				// 艦隊リスト更新用
				if let shipID = dock.valueForKey("ship_id") as? String {
					let shipPredicate = NSPredicate(format: "id = \(shipID)")
					error = nil
					let shipArray = store.objectsWithEntityName("Ship",
						sortDescriptors: nil,
						predicate: shipPredicate,
						error: &error)
					if shipArray.count == 0 {
						if error != nil {
							log("error -> \(error!)", argList: getVaList([]))
						}
						return
					}
					if let ship = shipArray[0] as? HMKCShipObject {
						ship.nowhp = ship.maxhp
					}
				}
			}
		}
	}
}
